### ISSUE TITLE
make optimisticLocking a store feature?

### ISSUE THREAD

**Issue by amcp** [thinkaurelius/titan#1135]
These tests override isLockingOptimistic:
https://github.com/thinkaurelius/titan/blob/47e9b35efaa4d1eb8722c1db0b1e599336a2c049/titan-cassandra/src/test/java/com/thinkaurelius/titan/graphdb/CassandraGraphTest.java#L28
https://github.com/thinkaurelius/titan/blob/47e9b35efaa4d1eb8722c1db0b1e599336a2c049/titan-berkeleyje/src/test/java/com/thinkaurelius/titan/graphdb/berkeleyje/BerkeleyGraphTest.java#L62
https://github.com/thinkaurelius/titan/blob/47e9b35efaa4d1eb8722c1db0b1e599336a2c049/titan-test/src/test/java/com/thinkaurelius/titan/graphdb/inmemory/InMemoryGraphTest.java#L71
https://github.com/thinkaurelius/titan/blob/47e9b35efaa4d1eb8722c1db0b1e599336a2c049/titan-hbase-parent/titan-hbase-core/src/test/java/com/thinkaurelius/titan/graphdb/hbase/HBaseGraphTest.java#L32
https://github.com/thinkaurelius/titan/blob/47e9b35efaa4d1eb8722c1db0b1e599336a2c049/titan-cassandra/src/test/java/com/thinkaurelius/titan/graphdb/thrift/ThriftGraphCacheTest.java#L25

This value varies from one storage backend to another. Those differences are summarized [here](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Tools.TitanDB.html#Tools.TitanDB.TitanFeatures), and I think it makes sense to pull this out of tests into StoreFeatures. If you agree I could make a pull request...

Thanks
Alex

**Comment by amcp** [comment]
created pull request https://github.com/amcp/titan/commit/0638987c7b5907a790e880e1eec305b49ac6494c

**Comment by dalaro** [comment]
`isLockingOptimistic` was a hack in the test code before the PR, but if we're pulling it into src/main like this, then perhaps we should refine the concept a bit.  Thinking out loud here...

It might be misleadingly named, since "pessimistic" behavior on BDB only really applies to its read locks; write lock acquisition is still deferred until the titan transaction commits.

This behavior is also not strictly a store feature; I suppose it's determined by the (locker, store) pair.   On the other hand, with the only non-default alternative locker deleted and likely to remain that way (#1131), we could probably get away with treating optimism as a strictly store-only property, even though it isn't theoretically true.  This could be a problem for someone writing a custom locker.

I see that optimisticLocking is listed as a table row in the linked documentation, but I don't understand exactly what contract this implies to the user.  Is it the same property we use in the test code -- telling us in what order conflicting transactions are expected to fail?

**Comment by amcp** [comment]
I agree the name is probably not right. It is the same property used in test code, determining the order in which conflicting transactions are expected to fail. If you can suggest a better name for this feature, I can update it in the pull request and in the linked documentation.

I will fix the imports in another pull request.

**Comment by amcp** [comment]
There may not be much wrong with the suggested name, as long as the documentation of the feature in StoreFeatures says that it works in conjunction with the Locker in place.

**Comment by dalaro** [comment]
I haven't thought about this carefully yet, but perhaps moving `isLockingOptimistic` (or whatever we call it) to `GraphDatabaseConfiguration` would obviate the hypothetical problem about considering the locker.  Its implementation could still just call and return some boolean-returning method on `StoreFeatures` right now, functionally equivalent to the PR.  However, exposing this publicly through GDC instead of StoreFeatures would let us account for locker behavior if we need to without affecting user code that just wants information about conflicting transaction failure order.  What do you think?

**Comment by amcp** [comment]
I think that exposing isLockingOptimistic in GraphDatabaseConfiguration would allow us to consider the locker. Let me revise the PR.

**Comment by amcp** [comment]
https://github.com/thinkaurelius/titan/pull/1140
